editor/gui/scene_tree_editor.cpp

```diff
@@ -238,34 +238,106 @@ void SceneTreeEditor::_add_nodes(Node *p_node, TreeItem *p_parent) {
 		part_of_subscene = p_node != get_scene_node() && get_scene_node()->get_scene_inherited_state().is_valid() && get_scene_node()->get_scene_inherited_state()->find_node_by_path(get_scene_node()->get_path_to(p_node)) >= 0;
 	}
 
-	TreeItem *item = tree->create_item(p_parent);
+	TreeItem *item;
+	int node_index = p_node->get_index(false);
+	HashMap<Node *, CachedNode>::Iterator I = node_cache.find(p_node);
+
+	if (I) {
+		item = I->value.item;
+	} else {
+		// We don't set the node_index cache here. As undo/redo might've just recreated this.
+		item = tree->create_item(p_parent);
+		CachedNode cached_node;
+		cached_node.item = item;
+		I = node_cache.insert(p_node, cached_node);
+	}
+
+	// Do a quick check to see if we are where we expect to be.
+	// We can't only do this when we're dirty as moving one node will necessarily move our siblings.
+	if (p_parent && I->value.index != node_index) {
+		TreeItem *new_neigbor = cast_to<TreeItem>(p_parent->get_child(node_index);
+
+		if (new_neigbor && new_neigbor != item) {
```

Comment:
As mentioned before, you need to check before `get_child()` whether the index is within `get_child_count()`. It's the only case when `new_neighbor` will be null, but it will also print an error.